store/types/store.go

```diff
@@ -0,0 +1,381 @@
+//
+// This is an optional, but useful extension to any CommitStore
+type Queryable interface {
+	Query(abci.RequestQuery) abci.ResponseQuery
+}
+
+//----------------------------------------
+// MultiStore
+
+type MultiStore interface { //nolint
+	Store
+
+	// Cache wrap MultiStore.
+	// NOTE: Caller should probably not call .Write() on each, but
+	// call CacheMultiStore.Write().
+	CacheMultiStore() CacheMultiStore
+
+	// Convenience for fetching substores.
+	// If the store does not exist, panics.
+	GetStore(StoreKey) Store
+	GetKVStore(StoreKey) KVStore
+
+	// TracingEnabled returns if tracing is enabled for the MultiStore.
+	TracingEnabled() bool
+
+	// WithTracer sets the tracer for the MultiStore that the underlying
+	// stores will utilize to trace operations. A MultiStore is returned.
+	WithTracer(w io.Writer) MultiStore
+
+	// WithTracingContext sets the tracing context for a MultiStore. It is
+	// implied that the caller should update the context when necessary between
+	// tracing operations. A MultiStore is returned.
+	WithTracingContext(TraceContext) MultiStore
+
+	// ResetTraceContext resets the current tracing context.
+	ResetTraceContext() MultiStore
+}
+
+// From MultiStore.CacheMultiStore()....
+type CacheMultiStore interface {
+	MultiStore
+	Write() // Writes operations to underlying KVStore
+}
+
+// A non-cache MultiStore.
+type CommitMultiStore interface {
+	Committer
+	MultiStore
+
+	// Mount a store of type using the given db.
+	// If db == nil, the new store will use the CommitMultiStore db.
+	MountStoreWithDB(key StoreKey, typ StoreType, db dbm.DB)
+
+	// Panics on a nil key.
+	GetCommitStore(key StoreKey) CommitStore
+
+	// Panics on a nil key.
+	GetCommitKVStore(key StoreKey) CommitKVStore
+
+	// Load the latest persisted version.  Called once after all
+	// calls to Mount*Store() are complete.
+	LoadLatestVersion() error
+
+	// Load a specific persisted version.  When you load an old
+	// version, or when the last commit attempt didn't complete,
+	// the next commit after loading must be idempotent (return the
+	// same commit id).  Otherwise the behavior is undefined.
+	LoadVersion(ver int64) error
+}
+
+//---------subsp-------------------------------
+// KVStore
+
+// KVStore is a simple interface to get/set data
+type KVStore interface {
+	Store
+
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get(key []byte) []byte
+
+	// Has checks if a key exists. Panics on nil key.
+	Has(key []byte) bool
+
+	// Set sets the key. Panics on nil key or value.
+	Set(key, value []byte)
+
+	// Delete deletes the key. Panics on nil key.
+	Delete(key []byte)
+
+	// Iterator over a domain of keys in ascending order. End is exclusive.
+	// Start must be less than end, or the Iterator is invalid.
+	// Iterator must be closed by caller.
+	// To iterate over entire domain, use store.Iterator(nil, nil)
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
```

Comment:
fwiw we don't obey this contract at *all* (assuming writes include deletions, but even so) - we break it all over the codebase

is this actually required?

Reply:
The iavl tree requires, but since we are cache wrapping the iavlstore before passing it to the handlers and the modification on the `CacheKVStore` is not reflected on the iterator, it looks safe to do it in the modules. It is still an exception but for the module writers' perspective we should update the comment to notify them that it is safe to modify in the modules

Reply:
Reasonable - let's definitely update the comment